File: src/modules/workshop/utils/naturalOrder.ts
/**
 * The name comparator every workshop file listing sorts by.
 *
 * Some listings sort here and some arrive sorted from the backend, so this
 * mirrors `compare_names` in `crates/ltk-manager-core/src/utils/natural_order.rs`.
 * Change one and the other's `naturalOrder.fixture.json` test fails.
 */

/** Whether a UTF-16 code unit is an ASCII digit. */
function isDigit(code: number): boolean {
  return code >= 0x30 && code <= 0x39;
}

/** Fold an ASCII upper-case code unit, leaving everything else alone. */
function foldCase(code: number): number {
  return code >= 0x41 && code <= 0x5a ? code + 0x20 : code;
}

function trimLeadingZeros(digits: string): string {
  let first = 0;
  while (first < digits.length && digits.charCodeAt(first) === 0x30) first += 1;
  return digits.slice(first);
}

/**
 * Order two file names the way a file explorer does.
 *
 * A run of digits compares as a number, so `skin9` precedes `skin50`. Letters
 * compare without regard to case. Names that differ only in the padding of a
 * number, or only in case, fall back to code unit order, which keeps this a
 * total order.
 *
 * `Intl.Collator` cannot stand in for this. It resolves to the host locale, so
 * two machines would disagree, and `sensitivity: "base"` calls names that differ
 * only in case equal, which leaves their order to however the sort found them.
 */
export function compareNames(a: string, b: string): number {
  let i = 0;
  let j = 0;

  while (i < a.length && j < b.length) {
    if (isDigit(a.charCodeAt(i)) && isDigit(b.charCodeAt(j))) {
      const aStart = i;
      const bStart = j;
      while (i < a.length && isDigit(a.charCodeAt(i))) i += 1;
      while (j < b.length && isDigit(b.charCodeAt(j))) j += 1;

      const aDigits = trimLeadingZeros(a.slice(aStart, i));
      const bDigits = trimLeadingZeros(b.slice(bStart, j));
      /* Longer is larger once the padding is off, so this reads the two runs as
      numbers without parsing one that would overflow. */
      if (aDigits.length !== bDigits.length) return aDigits.length - bDigits.length;
      if (aDigits !== bDigits) return aDigits < bDigits ? -1 : 1;
      continue;
    }

    const aFolded = foldCase(a.charCodeAt(i));
    const bFolded = foldCase(b.charCodeAt(j));
    if (aFolded !== bFolded) return aFolded - bFolded;
    i += 1;
    j += 1;
  }

  const remaining = a.length - i - (b.length - j);
  if (remaining !== 0) return remaining;
  return a < b ? -1 : a > b ? 1 : 0;
}

File: crates/ltk-manager-core/src/utils/natural_order.rs
//! Natural order over file names, matching what the frontend sorts by.
//!
//! The twin is `compareNames` in `src/modules/workshop/utils/naturalOrder.ts`.
//! Change one and the other's `naturalOrder.fixture.json` test fails.

use std::cmp::Ordering;

/// Compare two file names the way a file explorer does.
pub fn compare_names(a: &str, b: &str) -> Ordering {
    let (left, right) = (a.as_bytes(), b.as_bytes());
    let (mut i, mut j) = (0, 0);

    while i < left.len() && j < right.len() {
        if left[i].is_ascii_digit() && right[j].is_ascii_digit() {
            let (a_run, b_run) = (digit_run(&left[i..]), digit_run(&right[j..]));
            i += a_run.len();
            j += b_run.len();

            let (a_digits, b_digits) = (trim_leading_zeros(a_run), trim_leading_zeros(b_run));
            /* Longer is larger once the padding is off, so this reads the two
            runs as numbers without parsing one that would overflow. */
            match a_digits
                .len()
                .cmp(&b_digits.len())
                .then_with(|| a_digits.cmp(b_digits))
            {
                Ordering::Equal => continue,
                order => return order,
            }
        }

        match left[i]
            .to_ascii_lowercase()
            .cmp(&right[j].to_ascii_lowercase())
        {
            Ordering::Equal => {
                i += 1;
                j += 1;
            }
            order => return order,
        }
    }

    /* The length decides which name is the prefix of the other, and the byte
    compare only breaks ties the fold and the padding left equal. */
    (left.len() - i)
        .cmp(&(right.len() - j))
        .then_with(|| a.cmp(b))
}

fn digit_run(bytes: &[u8]) -> &[u8] {
    let end = bytes
        .iter()
        .position(|byte| !byte.is_ascii_digit())
        .unwrap_or(bytes.len());
    &bytes[..end]
}

fn trim_leading_zeros(digits: &[u8]) -> &[u8] {
    let first = digits
        .iter()
        .position(|&byte| byte != b'0')
        .unwrap_or(digits.len());
    &digits[first..]
}

#[cfg(test)]
mod tests {
    use super::*;

    /// The orderings the frontend's `compareNames` must produce too.
    ///
    /// Read `naturalOrder.fixture.json`'s own comment for the contract.
    #[test]
    fn obeys_the_shared_ordering_fixture() {
        #[derive(serde::Deserialize)]
        struct Fixture {
            cases: Vec<Case>,
        }
        #[derive(serde::Deserialize)]
        struct Case {
            name: String,
            input: Vec<String>,
            expect: Vec<String>,
        }

        let raw = include_str!(
            "../../../../src/modules/workshop/utils/__tests__/naturalOrder.fixture.json"
        );
        let fixture: Fixture = serde_json::from_str(raw).unwrap();

        for case in fixture.cases {
            let mut sorted = case.input.clone();
            sorted.sort_by(|a, b| compare_names(a, b));
            assert_eq!(sorted, case.expect, "{}", case.name);

            /* Reversed too, so a case the given order already satisfies does not
            pass for that reason. */
            let mut reversed: Vec<String> = case.input.into_iter().rev().collect();
            reversed.sort_by(|a, b| compare_names(a, b));
            assert_eq!(reversed, case.expect, "{} (reversed)", case.name);
        }
    }

    #[test]
    fn equal_names_compare_equal() {
        assert_eq!(compare_names("skin1.bin", "skin1.bin"), Ordering::Equal);
    }
}
